schema/schema/generate_schema.go

```diff
@@ -0,0 +1,267 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schema
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"reflect"
+	"strings"
+
+	"knative.dev/pkg/schema/docs"
+)
+
+var (
+	trueVal = true
+
+	zero      = 0.0
+	int8Max   = float64(math.MaxInt8)
+	int16Max  = float64(math.MaxInt16)
+	uint8Max  = float64(math.MaxUint8)
+	uint16Max = float64(math.MaxUint16)
+	uint32Max = float64(math.MaxUint32)
+)
+
+func GenerateForType(t reflect.Type) JSONSchemaProps {
+	s := generateStructSchema(t, true)
+	return s
+}
+
+func generateSchema(t reflect.Type, history ...reflect.Type) JSONSchemaProps {
+	if visited(t, history) {
+		return JSONSchemaProps{
+			Type: "object",
+		}
+	}
+
+	switch k := t.Kind(); k {
+	case reflect.Bool:
+		return JSONSchemaProps{
+			Type: "boolean",
+		}
+	case reflect.Int:
+		return JSONSchemaProps{
+			Type:   "integer",
+			Format: "int32",
+		}
+	case reflect.Int8:
+		return JSONSchemaProps{
+			Type:    "integer",
+			Maximum: &int8Max,
+		}
+	case reflect.Int16:
+		return JSONSchemaProps{
+			Type:    "integer",
+			Maximum: &int16Max,
+		}
+	case reflect.Int32:
+		return JSONSchemaProps{
+			Type:   "integer",
+			Format: "int32",
+		}
+	case reflect.Int64:
+		return JSONSchemaProps{
+			Type:   "integer",
+			Format: "int64",
+		}
+	case reflect.Uint:
+		return JSONSchemaProps{
+			Type:    "integer",
+			Minimum: &zero,
+		}
+	case reflect.Uint8:
+		return JSONSchemaProps{
+			Type:    "integer",
+			Minimum: &zero,
+			Maximum: &uint8Max,
+		}
+	case reflect.Uint16:
+		return JSONSchemaProps{
+			Type:    "integer",
+			Minimum: &zero,
+			Maximum: &uint16Max,
+		}
+	case reflect.Uint32:
+		return JSONSchemaProps{
+			Type:    "integer",
+			Format:  "int64",
+			Minimum: &zero,
+			Maximum: &uint32Max,
+		}
+	case reflect.Uint64:
+		return JSONSchemaProps{
+			Type:    "integer",
+			Format:  "int64",
+			Minimum: &zero,
+		}
+
+	case reflect.Float32:
+		return JSONSchemaProps{
+			Type:   "number",
+			Format: "float",
+		}
+	case reflect.Float64:
+		return JSONSchemaProps{
+			Type:   "number",
+			Format: "double",
+		}
+	case reflect.Map:
+		s := generateMapSchema(t)
+		return s
+	case reflect.Ptr:
+		return generateSchema(t.Elem(), history...)
+	case reflect.Slice:
+		s := generateSliceSchema(t)
+		return s
+	case reflect.String:
+		return JSONSchemaProps{
+			Type: "string",
+		}
+	case reflect.Struct:
+		history = append(history, t)
+		s := generateStructSchema(t, false, history...)
+		return s
+	case reflect.Interface:
+		return JSONSchemaProps{
+			Type: "object",
+		}
+	case reflect.Uintptr:
+		fallthrough
+	case reflect.Complex64:
+		fallthrough
+	case reflect.Complex128:
+		fallthrough
+	case reflect.Array:
+		fallthrough
+	case reflect.Chan:
+		fallthrough
+	case reflect.Func:
+		fallthrough
+	case reflect.UnsafePointer:
+		panic(fmt.Errorf("can't handle kind %+v", t))
+	default:
+		panic(fmt.Errorf("unknown kind: %+v", t))
```

Comment:
Did you mean `%T` here, and in the previous case?

Reply:
t is already a reflected Type so `%T`  I think would spit out reflection.Type

Reply:
Ah right, fair point.